Legger til støtte for å journalføre dokument på virksomhet.

janitamarielle opened the pull request:
Legger til støtte for å journalføre dokument på virksomhet.
- Man kan velge mellom å journalføre på bruker og virksomhet. Dette har endel følger lengre nedover i journalføringsbildet.
- Man kan nå hente fagsaker basert på orgnr i tillegg til ident og saksnummer.


https://jira.adeo.no/browse/MELOSYS-4896

Melosys-api: https://github.com/navikt/melosys-api/pull/1122 ✅ 
Melosys-docker-compose: https://github.com/navikt/melosys-docker-compose/pull/38 ✅

janitamarielle committed:
Legger til støtte for å journalføre dokument på virksomhet.
- Man kan velge mellom å journalføre på bruker og virksomhet. Dette har endel følger lengre nedover i journalføringsbildet.
- Man kan nå hente fagsaker basert på orgnr i tillegg til ident og saksnummer.

IsaSalemHame commented on src/sider/journalforing/komponenter/journalforingSchema.js:
```diff
@@ -36,35 +36,44 @@ const kreverPeriode = (journalforingHensikt, behandlingstema) =>
-  avsenderID: string().when(["avsenderType", "$erAvsenderPreutfylt"], {
-    is: organisasjonOgIkkePreutfyltAvsender,
+  virksomhetNavn: string().nullable(),
+  avsenderID: string().when(["avsenderType", "$erAvsenderPreutfylt", "journalføresPå"], {
```
Avsender bør resettes når man bytter mellom Bruker og virksomhet. Dersom man f.eks. begynner å journalføre på person, velger f.eks. `Utenlandsk trygdemyndighet`, for så å gå til Virksomhet og ikke velge hverken "Virksomhet" eller "Fullmektig", så vil det fortsatt gå an å journalføre

janitamarielle replied:
Done

IsaSalemHame commented on src/sider/journalforing/komponenter/informasjon.js:
```diff
@@ -165,33 +200,78 @@ class Informasjon extends Component {
+            </span>
+          )}
+          {visVirksomhetSpinner && <Nav.NavFrontendSpinner className="informasjon__spinner" />}
+        </>
```
Vi skulle ikke hatt litt mer spacing mellom her og "Informasjon om avsender"? Ligger veldig tett føler jeg. Mulig det var sånn før også, men virker å være mer spacing i skissene.

janitamarielle replied:
Det kommer nok i skisse-saken nevnt tidligere.

IsaSalemHame commented on src/sider/journalforing/komponenter/informasjon.js:
```diff
@@ -165,33 +200,78 @@ class Informasjon extends Component {
+          }
+        </FeatureToggle>
 
         <Mui.Undertittel tekst="Informasjon om avsender" ikon={Ikoner.Globe} className="undertittel" />
```
Det er annerledes ikon i skissene. Samt. er det understrek under hver overskrift. Skulle vi kanskje fått slike designendringer i egen PR, hvis det ikke har kommet egen task på det?

IsaSalemHame commented on src/sider/journalforing/komponenter/avsendere.js:
```diff
@@ -13,13 +13,16 @@ import "./avsendere.css";
 export const AvsenderOrganisasjon = ({
```
Annerledes design mellom her og skissene. Igjen, egen task, eller her?

janitamarielle replied:
Det er laget en egen task: https://jira.adeo.no/browse/MELOSYS-5091

Tar skissene på ting som er nytt i denne tasken, men fokuserer ellers på funksjonalitet :D

IsaSalemHame commented on src/sider/journalforing/journalforing.js:
```diff
@@ -99,12 +101,14 @@ class Journalforing extends Component {
       avsenderNavn,
-      brukerID,
+      brukerID: journalføresPåBruker ? brukerID : null,
+      virksomhetOrgnr: !journalføresPåBruker ? virksomhetOrgnr : null,
```
Trenger man å ha if-setting her? Er ikke verdien null hvis ikke det eksisterer?

janitamarielle replied:
Hmm, det er mest fordi brukerID/virksomhetOrgnr blir resatt til "" og ikke null. Men kan se hvor lett det er å endre til null og da slippe denne.

janitamarielle replied:
Fixed ✅

IsaSalemHame commented on src/sider/journalforing/journalforing.js:
```diff
@@ -302,10 +307,30 @@ class Journalforing extends Component {
+    if (!response || !response.data) {
+      return false;
+    }
+    const { navn = "" } = response.data;
```
Er det egentlig noe vits med `= ""`? Bør ikke ha noe å si for if-sjekken, såvidt jeg kan se.

janitamarielle replied:
Dette var mest for å få det konsekvent med hentOgVisBruker. Endrer til:
```
    const response = await sokOrgnr(virksomhetOrgnr);
    const navn = response?.data?.navn;
    if (Utils._isEmpty(navn)) {
      return false;
    }
```

IsaSalemHame commented on src/sider/journalforing/journalforing.js:
```diff
@@ -302,10 +307,30 @@ class Journalforing extends Component {
+      return false;
+    }
+    settFeltInnhold("virksomhetNavn", navn);
+    await hentFagsakListeForOrgnr(virksomhetOrgnr);
```
Lurer på om hentFagsakListeForOrgnr og hentFagsakListeForFnr burde ut av disse metodene. Slik jeg skjønner, så går denne videre til JournalføringForm, og videre inn i Informasjon.js (som forsåvidt jeg synes er litt dårlig navn). Jeg mener at den som kaller på metoden burde få en promise, som så kaller på hentFagsakListeForOrgnr.

janitamarielle replied:
Skrev om nå. Syns det ble mye mer oversiktlig nå samtidig som man fikk ryddet opp i mye kode som ikke var i bruk

IsaSalemHame replied:
nice

IsaSalemHame commented on src/sider/journalforing/komponenter/avsendere.js:
```diff
@@ -79,7 +83,12 @@ AvsenderOrganisasjon.defaultProps = {
 };
 
-export const AvsenderFullmektig = ({ avsenderID, settFeltInnhold, hentOgVisRepresentant }) => {
+export const AvsenderFullmektig = ({
```
Forskjellige avsendere burde egentlig bare ha egne filer, kanskje i egen mappe med en index.js. Litt vanskelig å vite at det finnes flere komponenter i samme fil..

janitamarielle replied:
Kan kanskje lage en avsendere-folder under /komponenter og skille ut AvsenderOrganisasjon, AvsenderFullmektig og AvsenderUtenlandskTrygdemyndighet.

@martonSkjevelandNav Ødelegger det for deg dersom dette gjøres her?

janitamarielle replied:
Ser ut som marton har fiksa i https://github.com/navikt/melosys-web/pull/1484/files 👏

IsaSalemHame commented on src/sider/journalforing/komponenter/avsendervelger.js:
```diff
@@ -68,63 +83,55 @@ const AvsenderVelger = ({
+              value={KV.AvsenderTyper.ARBEIDSGIVER}
+              className="avsendervelger__radio"
+            />
+            {formValues.avsenderType === KV.AvsenderTyper.ARBEIDSGIVER && (
```
Hvis du bare refresher siden, og trykker på denne, så kommer feil i console:
`Warning: Failed prop type: The prop `journalføresPåVirksomhet` is marked as required in `AvsenderOrganisasjon`, but its value is `undefined`.`

janitamarielle replied:
Løst i https://github.com/navikt/melosys-web/pull/1458/commits/a0dd426ccc62501802dc48703da6f0c7db0bd511. Var jeg som hadde glemt å legge til props i alle type avsendere 🙃

IsaSalemHame commented on src/sider/journalforing/komponenter/komponent.tsx:
```diff
@@ -0,0 +1,24 @@
+  innhold: ReactElement;
+}
+
+const Komponent = ({ ikon: Ikon, tittel, innhold }: KomponentProps) => {
```
Er det ikke bedre å kalle det JournalføringKomponent? Da er det litt lettere å se hva "Komponent" er når man randomly er inne på denne komponenten.

janitamarielle replied:
Hmm, jeg kan det, men den er jo allerede innenfor journalforing-mappen. 
Blir kanskje mye duplisering og vanskelig å finne frem dersom alt skal ha prefiks journalføring? 

Denne komponenten er ment kun brukt for journalføring så man skal egentlig ikke randomly komme inn 🤔

frans-a replied:
Tenker litt som Janita her. Har vi ikke nok kontekst allerede Isa?
Samtidig kaller man f.eks. i api `FagsakService` selv om klassen finnes under pakken fagsak

IsaSalemHame replied:
Delvis enig med dere, men i dette tilfellet er det omtrent 20 andre filer i mappen for "journalforing", og "komponent" er såpass generisk (og litt reservert) navn at det er misvisende hva det representerer.

Da jeg skrev den kommentaren trodde jeg at det var "hovedkomponenten" for journalføring. Når jeg ser nærmere på det, så er det egentlig et felt det representerer. In any case så mener jeg det er dårlig navn. 

Ved nærmere kikk så er jeg i tvil på om hele komponenten er i det hele tatt nødvendig. Det blir brukt ett sted, og gjør ikke at man sparer kode elns.

janitamarielle replied:
> Ved nærmere kikk så er jeg i tvil på om hele komponenten er i det hele tatt nødvendig. Det blir brukt ett sted, og gjør ikke at man sparer kode elns.

Vet ikke om det kom frem godt nok, men denne er tenkt brukt i alle delene/komponentene i journalføringsbildet siden det er nye skisser på journalføring. Siden det blir for mye å ta all styling/skisse-endring nå er det laget egen jira-sak på det.

IsaSalemHame commented on src/sider/journalforing/komponenter/dokumentetJournalføresPå.tsx:
```diff
@@ -0,0 +1,68 @@
+
+type PropsFromRedux = ConnectedProps<typeof connector>;
+
+const DokumentetJournalføresPå = ({ journalføresPå, oppdaterFelt, defaultBrukerID }: PropsFromRedux) => {
```
Tror ikke vi har noen andre filer med æ ø å. Burde kanskje fortsette med det, bare for å være konsistent. Vet ikke om det er derfor, men ser faktisk ut som det ødelegger for Pull Requests i IntelliJ.

Skulle helst ha kalt det noe annet også, kanskje "JournalføringRolleVelger" eller noe sånt. Føler komponenten som det heter nå er mer en setning enn et komponentnavn.

janitamarielle replied:
Skal endre når vi bestemmer oss for navn. Tenkte vi hadde gått for å bruker ÆØÅ, men dersom det skaper issues i git må vi jo unngå det.

frans-a replied:
Vi har brukt ÆØÅ i filer uten problemer, men i filnavn er det fortsatt problematisk og bør unngås

IsaSalemHame commented on src/kodeverk/koder.ts:
```diff
@@ -161,6 +161,11 @@ export const MottakerRolle = {
   REPRESENTANT: "REPRESENTANT",
 };
 
+export const JournalføringRolle = {
```
Rolle eller Aktør? Idk selv. Men spørsmål du kan tenke på.

janitamarielle replied:
Tror aktør kan være misvisende mtp aktørId og at vi egentlig kun bruker det når det er snakk om bruker 🤔

frans-a replied:
Bør det defineres her? Alle konstanter er ikke nødvendigvis kodeverk, dette er ikke offisielt kodverk som https://kodeverk-web.dev.adeo.no eller https://confluence.adeo.no/display/TEESSI/Kodeverk+i+Melosys. Det kunne defineres nærmere journalføring tenkte jeg.

Disse konstanter definerer hvem gjelder forsendelsen og journalføringen. Det kunne kanskje kalles `JournalføringGjelder`? Både `JournalføringRolle` og aktør er utydelig for meg

janitamarielle replied:
Enig, disse kan legges i journalførings-komponenten 🤔 

Mina likte `JournalføringGjelder` så da går vi for den 👍

IsaSalemHame replied:
JournalføringGjelder hørtes mer tydelig ut.

IsaSalemHame commented on src/sider/journalforing/komponenter/journalforingSchema.js:
```diff
@@ -36,35 +36,44 @@ const kreverPeriode = (journalforingHensikt, behandlingstema) =>
-      then: string()
-        .harIkkeOrgnrLengde(FANT_INGEN_NAVN_PA_ORGNR)
-        .harIkkeFnrEllerDnrLengde(FANT_INGEN_NAVN_PA_FNR_ELLER_DNR),
+    .when("journalføresPå", {
```
Skulle man heller ha kalt det "journalføringsrolle" / "journalføringsaktør" i steden for "journalføresPå"?

janitamarielle replied:
Hører med Mina. Skal grubles litt på 🤔 ⏳

janitamarielle replied:
Vi prøver JournalføringGjelder

IsaSalemHame replied:
JournalføringGjelder hørtes bra ut

janitamarielle committed:
Merge branch 'master' of https://github.com/navikt/melosys-web into MELOSYS-4896_journalfør-dokument-på-virksomhet

 Conflicts:
	src/sider/journalforing/journalforing.js

janitamarielle committed:
Setter tom-verdien til null istedenfor "" innenfor journalforing-domenet. Hovedsaklig på schema-feltene: bruker, virksomhet, representant og avsender.

janitamarielle committed:
Resetter avsender-felt ved endring av hvem dokumentet journalføres på (bruker/virksomhet)

janitamarielle committed:
Unødvendig logikk på avsendertype

janitamarielle committed:
Merge branch 'master' of https://github.com/navikt/melosys-web into MELOSYS-4896_journalfør-dokument-på-virksomhet

frans-a commented on src/sider/journalforing/komponenter/avsendere.js:
```diff
@@ -13,16 +13,19 @@ import "./avsendere.css";
     if (avsenderType === KV.AvsenderTyper.ARBEIDSGIVER_FULLMEKTIG) {
       settFeltInnhold("representantRepresenterer", MKV.Koder.representerer.BRUKER);
+    } else if (journalføresPåVirksomhet) {
+      settFeltInnhold("representantRepresenterer", MKV.Koder.representerer.ARBEIDSGIVER);
```
Dette ser ikke ut til å stemme. Hvis det journalføres på virksomhet blir fullmektig representant for virksomheten, ikke arbeidsgiver.

> det kun er mulig at fullmektig er fullmektig for "Virksomhet"

janitamarielle replied:
Hmm, dette stemmer. Den representerer jo virksomheten, som i dette tilfellet er BRUKER... 🤔 

Skal vi fortsette med representerer-kode eller potensielt rename den BRUKER_ELLER_VIRKSOMHET? Vil tro det er generelt samme oppførsel på BRUKER og VIRKSOMHET i dette tilfellet, og at man derfor ikke trenger en helt ny kode. Men her er jeg usikker.

frans-a replied:
:warning: bruker i Melosys-kontekst er ikke det samme som bruker i Joark-kontekst.
Bruker i Melosys betyr personen vi prøver å avklare lovvalg og trygdetilhørighet for.
Mens bruker hos Joark betyr bare hvem gjelder dokumentet som skal journalføres.

Så siden representant er i Melosys-kontekst blir det en forskjell om vi lagrer representant som representerer en person (bruker) i stedet for fullmektig.

Se https://github.com/navikt/melosys-api/pull/1122#issuecomment-1114775530

janitamarielle replied:
Enig i at vi ikke kan bruke Representerer.BRUKER. Spørsmålet er om vi skal lage ny Representerer.VIRKSOMHET eller om vi skal rename Representerer.BRUKER_ELLER_VIRKSOMHET 🤔 

Det kommer jo an på om melosys skal behandle dem forskjellig.

frans-a replied:
Representerer.BRUKER brukes allerede, ja, og kan ikke ikke bety at virksomhet representeres.
Jeg ser ikke helt hvordan BRUKER_ELLER_VIRKSOMHET skulle fungere.
Om man representerer ARBEIDSGIVER eller VIRKSOMHET er det ikke nødvendigvis det samme.

janitamarielle replied:
Hmm, for nå har vi BRUKER, ARBEIDSGIVER, og BEGGE i kodeverket "representerer". 

Til nå har det betydd at skal bruker ha brev sender man det til den fullmektige for bruker. 
Det vil vel være tilsvarende logikk for fullmektig for virksomhet: Dersom virksomheten skal ha brev sender man det istedenfor til den fullmektige. 
Syns altså at det er lik logikk for fullmektig som representerer BRUKER og en mulig ny VIRKSOMHET-kode. Hvorfor ikke da sette BRUKER til BRUKER_ELLER_VIRKSOMHET dersom logikken kommer til å være den samme?

Bruker har frem til nå vært den man journalfører på, eller den nye journalføringenGjelder-feltet. Man er altså fullmektig for den journalføringen gjelder. Som nå kan være BRUKER eller VIRKSOMHET.

janitamarielle replied:
Skal jeg kalle inn møte med Yvonne (og Mina) om dette?

janitamarielle replied:
https://confluence.adeo.no/display/TEESSI/Kodeverk+i+Melosys#KodeverkiMelosys-Akt%C3%B8rsrollerogRepresenterer

~~Lager ny kode i kodeverket.~~ 
Nope, leste feil - så på aktør. Spør i melosys-fag.

janitamarielle replied:
[Vi skal ikke støtte fullmektig for virksomhet](https://github.com/navikt/melosys-api/pull/1122#issuecomment-1117139034)

janitamarielle committed:
Flytter hentOgVisBruker, hentOgVisVirksomhet, hentOgVisAvsender og hentOgVisRepresentant til informasjon siden det er der de er brukt.
Legger logikken til sjekkBruker, sjekkVirksomhet og sjekkAvsender inn i nevnte funksjoner.
Fjerner spinner siden de bare gir misvisende inntrykk til saksbehandler (og det ikke har fungert i prod på lenge).

IsaSalemHame commented on src/sider/journalforing/komponenter/komponent.tsx:
```diff
@@ -0,0 +1,24 @@
+        <Ikon className="ikon" />
+        <Nav.Typo.Undertittel className="tittel">{tittel}</Nav.Typo.Undertittel>
+      </div>
+      {innhold}
```
Vet ikke hva som er standard i React/hva som er best practices når det er ett child element, men kunne kanksje ha hatt {props.content} her, og ellers `<Komponent ...> <innholdet/> </Komponent>` når vi skal bruke den.

janitamarielle replied:
Her har jeg egentlig bare brukt det vi ellers har i prosjektet 🤔 Kan se hvordan det blir seende ut?

janitamarielle replied:
On 2. thought, tar gjerne å skipper denne dersom du ikke brenner for det.

IsaSalemHame replied:
Brenner ikke for det.

IsaSalemHame replied:
![image](https://user-images.githubusercontent.com/7471428/166241101-31a05cd8-6825-4086-9f47-7ece510a3c6a.png)

IsaSalemHame commented on src/sider/journalforing/komponenter/journalforingSchema.js:
```diff
@@ -36,53 +36,69 @@ const kreverPeriode = (journalforingHensikt, behandlingstema) =>
+  return organisasjonAliasTyper.includes(avsenderType) && (!erAvsenderPreutfylt || journalføresPå === VIRKSOMHET);
 };
 
 const journalforing = object().shape({
```
Ser ut som journalføringsknappen ikke lenger gjør noe

janitamarielle replied:
Skal sjekke. Vet jeg har gjort endringer i skjema uten lokaltesting

janitamarielle replied:
El problemo: https://github.com/navikt/melosys-web/pull/1458/commits/98b026cda43e06c0ee25923133144e10a8ce4ced

IsaSalemHame replied:
olala :smirk:

IsaSalemHame commented on src/sider/journalforing/komponenter/avsendervelger.js:
```diff
@@ -25,6 +26,10 @@ const AvsenderVelger = ({
         break;
       }
+      case KV.AvsenderTyper.VIRKSOMHET: {
+        kopierVirksomhetTilAvsender();
```
Hva er det denne gjør egentlig? Tømmer avsender/org.navn elns?

janitamarielle replied:
Denne setter vel avsender id og navn til virksomhet id og navn.

janitamarielle committed:
Trenger ikke kodeverk for journalføringsroller - bruker heller const som ligger der de blir brukt.
Renamer journalføresPå -> journalforingGjelder, og dokumenterJournalføresPå -> journalforingGjelder.

IsaSalemHame reviewed (commented):
Veldig godt jobbet!

janitamarielle committed:
Feil i setting av virksomhetErAvsender

janitamarielle committed:
Merge branch 'master' of https://github.com/navikt/melosys-web into MELOSYS-4896_journalfør-dokument-på-virksomhet

 Conflicts:
	src/sider/journalforing/komponenter/informasjon.js

janitamarielle committed:
Feil navn på test-props

janitamarielle committed:
Virksomhet kan ikke velge avsender - den er alltid virksomhet.

janitamarielle committed:
Merge branch 'master' of https://github.com/navikt/melosys-web into MELOSYS-4896_journalfør-dokument-på-virksomhet

janitamarielle commented:
Merk at jeg revertet de fleste endringene på avsender-domenet her. Halvveis på grunn av at VIRKSOMHET skal ha default avsender så man viser ikke disse lengre, og halvveis fordi den oppryddingen jeg har gjort også kommer i PR-en: https://github.com/navikt/melosys-web/pull/1484

IsaSalemHame reviewed (approved):
Very nice

janitamarielle committed:
Feil i eksisterende funksjonalitet

janitamarielle committed:
Merge branch 'master' of https://github.com/navikt/melosys-web into MELOSYS-4896_journalfør-dokument-på-virksomhet

 Conflicts:
	src/sider/journalforing/journalforing.js
	src/sider/journalforing/komponenter/avsendere.js
	src/sider/journalforing/komponenter/informasjon.js
	src/sider/journalforing/komponenter/journalforingSchema.js

janitamarielle committed:
Endringer etter merge fra master. Litt merge conflict, litt fjerning av avsenderTyper som ikke lengre brukes

janitamarielle committed:
Merge branch 'master' of https://github.com/navikt/melosys-web into MELOSYS-4896_journalfør-dokument-på-virksomhet

 Conflicts:
	src/sider/journalforing/komponenter/journalforingform.js

janitamarielle commented:
Ble ganske store endringer etter merge inn master, greit å få andre enn mine øyne på det 😄

janitamarielle committed:
Endringer etter merge fra master

IsaSalemHame reviewed (approved):
Har testet med kombinasjonene jeg kan komme på, og det ser ut til å funke som det skal. 
Det er to ting jeg la merke til, men som ikke har så mye med det du har gjort å gjøre. Nummer 2 er noe som kaaanskje bør gjøres her? Du kan sjekke og avgjøre hvor kritisk det er.

1. Etter å ha journalført kommer ikke navnet på organisasjonen opp på forsiden i oppgavelisten. Blir feks "Trygdeavtale - null - null". Bør fikses i egen oppgave, ettersom det involverer både frontend og backend.
2. Journalføring-knappen er tilgjengelig selv om man ikke har oppgitt orgnr/persnr.

janitamarielle committed:
journalforingGjelder kunne ikke sjekkes inline mot const. Skrev det ut i arrow-funksjon

janitamarielle commented:
> 1. Etter å ha journalført kommer ikke navnet på organisasjonen opp på forsiden i oppgavelisten. Blir feks "Trygdeavtale - null - null". Bør fikses i egen oppgave, ettersom det involverer både frontend og backend.

Mina har avklart. Det kommer i egen oppgave med andre endringer som må justeres for virksomhet. 

> 2. Journalføring-knappen er tilgjengelig selv om man ikke har oppgitt orgnr/persnr.

Knapp skal være tilgjengelig, men nå kommer feilmeldingen opp sammen med de andre når du trykker på den: https://github.com/navikt/melosys-web/pull/1458/commits/5b6eff84bb9796c8fd2cea7333722fb0a5a9bfa0

janitamarielle committed:
Ble flere like arrow-funksjoner. Lager egne funksjoner
